Getting started guide for the host firewall

pchaigno opened the pull request:
First version of the host firewall getting started guide (GSG). I still want to extend it with a more extensive troubleshooting section than we have in [`Network Policy > Host Policies`](https://docs.cilium.io/en/latest/policy/language/#host-policies-beta) (or I'll add a section in [`Troubleshooting`](https://docs.cilium.io/en/latest/troubleshooting/) for the host firewall). I may also extend the GSG later to include some example of policies on path hostns <-> pods.

Fixes: #12278
Updates: #11799
```release-note
Add a getting started guide for the host firewall
```

pchaigno committed:
docs: Getting started guide for host firewall

First version of the host firewall GSG. I still want to extend it with a
more extensive troubleshooting section than we have in Network Policy >
Host Policies (or I'll add a section in Troubleshooting for the host
firewall). I may also extend the GSG later to include some example of
policies on path hostns <-> pods.

Fixes: #12278
Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
docs: Remove beta from Host Policies title

We need to remove it from the title or otherwise it is included wherever
we reference host policies with:

    :ref:`HostPolicies`

There is already a note on the beta status below the title anyway.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
examples: Remove old policies/l3/host directory

Host examples are now hosted in examples/policies/host.

Signed-off-by: Paul Chaignon <paul@cilium.io>

pchaigno committed:
examples: Allow remote nodes in host policy example

To be on the safe side and avoid breaking the cluster, we can allow
remote nodes and the health endpoint. Without this, we will at least
drop some ICMP probes.

Signed-off-by: Paul Chaignon <paul@cilium.io>

joestringer commented on Documentation/gettingstarted/host-firewall.rst:
```diff
@@ -0,0 +1,156 @@
+
+When applying `HostPolicies` carelessly, it's easy to block the access to the
+nodes or to break the cluster's normal behavior (ex. by blocking communication
+with kube-api).
```
Maybe we can start with a more neutral start point than going straight to assuming careless policy apply behaviour ;-)

Here's a suggestion for alternative wording, feel free to take/modify for what you think makes sense:

```suggestion
`HostPolicies` enforce strict access control over connectivity to nodes. Particular care must be taken to ensure that when host policies are imported, Cilium does not block access to the nodes or break the cluster's normal behavior (for example by blocking communication with ``kube-apiserver``).
```

pchaigno replied:
Hehe, yep. I took your recommendation but dropped `strict` and extended `to nodes` to `to and from nodes`.

joestringer commented on Documentation/gettingstarted/host-firewall.rst:
```diff
@@ -0,0 +1,156 @@
+with kube-api).
+
+To avoid such issues, we can switch the host firewall in audit mode, to
+validate the impact of host policies before enforcing them.
```
Probably worth making it clear that policy audit mode disables policy enforcement, with similar wording to the policy GSG.

```suggestion
To avoid such issues, we can switch the host firewall in audit mode, to
validate the impact of host policies before enforcing them. When Policy Audit Mode is enabled, no network policy is enforced so this setting is *not recommended for production deployment*.
```

(I appreciate that we're probably moving folks from no security -> hostfw so it's not such a big deal. I also note that the instructions here use the `endpoint config` approach to individually enable policy audit for the host which mitigates the risk to policy enforcement for pods. So if you want to be less loud with the warning I could understand, above is suggestion taken from the other doc).

pchaigno replied:
I think that's light enough, I'll add it.

And yes, I took some time to write this because I wanted to be able to switch into policy audit mode only for the host firewall. Otherwise, we can't expect users to take this approach in more realistic environments. The downside is that it exposes `cilium endpoint config` a bit more than we've used to in the past.

joestringer commented on Documentation/gettingstarted/host-firewall.rst:
```diff
@@ -0,0 +1,156 @@
+
+::
+
+    $ HOST_EP_ID=$(cilium endpoint list -o json | jq '.[] | select( .status.identity.id == 1 ).id')
```
FWIW `jq` is not available in the Cilium container so this is not necessarily going to 'just work' for the average user.

pchaigno replied:
Ok, I think that's a big enough problem that we should find a way to address it. I want users to be able to copy paste those commands as much as possible.

One idea is to restrict the whole guide to a single node. I would ask users to set a label on a node and I'd apply the policy only for that label. Then, we can `jq` the Cilium pod for that node and use it in subsequent commands. So something like:
```bash
export NODE_NAME=k8s1
kubectl label node $NODE_NAME access=ssh
[...]
CILIUM_POD_NAME=$(kubectl -n $CILIUM_NAMESPACE get pods -o json | jq -r '.items[] | select( .metadata.labels."k8s-app" == "cilium" and .spec.nodeName == env.NODE_NAME ).metadata.name')
HOST_EP_ID=$(kubectl -n $CILIUM_NAMESPACE exec $CILIUM_POD_NAME -- cilium endpoint list -o json | jq '.[] | select( .status.identity.id == 1 ).id')
[...]
```
One downside is that it could be a bit harder for users to understand what's happening, as in what info. is `jq` grabbing. So I at least need to be cautious with variable names I use :-)

What do you think?

qmonnet replied:
Can't you simply replace `-o json | jq` by `-o jsonpath` with the relevant filter to get the endpoint id?

pchaigno replied:
Aaah, that's what `-o jsonpath` is!

joestringer commented on Documentation/gettingstarted/host-firewall.rst:
```diff
@@ -0,0 +1,156 @@
+    for the cluster to work properly are denied. They should appear as ``action
+    allow``.
+
+
```
We lack the actual instructions for the user on how to turn this monitor output into a rule here.

A couple of suggestions, either:
* Refer the user to https://docs.cilium.io/en/v1.8/gettingstarted/policy-creation/#observe-policy-verdicts
* Refactor the above doc into a generic enough piece that it could be reused for the host endpoint.
* Introduce some hostfw-specific instructions here.

pchaigno replied:
I'll refer users to `Creating policies from verdicts` (btw, we're not consistent in our title capitalization across the docs).

There's one bit that's probably more common with the host firewall, without being specific to the host firewall: its policies tend to rely more on entities, mostly because there's a lot more sources and destinations to allow. Maybe it would make sense to extend `Creating policies from verdicts` with one example of an entity-based rule?

joestringer replied:
I guess I'd prefer to put such logic here to keep the creating policies GSG as simple as possible, but if there's a minimal recommendation to make either in that creating policies GSG or in this doc here then sure. It probably makes sense to at least mention that entities policies can be useful for enforcement(+link). This isn't a panacea, we won't cover all possible options, ultimately we still leave it up to the user to understand the policy model / what's available and craft their policies using the constructs we have available.

joestringer reviewed (changes requested):
Awesome guide, thanks.

I didn't try it out, but I noticed some issues that may cause problems for users. I also gave some wordsmithing suggestions below.

pchaigno committed:
docs: Use node labels in host firewall GSG

Relying on node labels to apply the host policy in the host firewall
getting started guide allows us to include fully working commands
throughout the guide.

Without this, commands cannot include the 'kubectl exec' part because we
don't know on what node and cilium pod the user is executing them.

Signed-off-by: Paul Chaignon <paul@cilium.io>

joestringer reviewed (approved):
Looks good, didn't try it out yet.
